JSON flag

bmiles opened the pull request:
Adds a `--json` flag to the CLI commands `projects`, `runs`, `protocols`.

Adds some usability when piping.

### Example:
```
> transctiptic protocols --remote --json | jq '.[0].inputs'
{
  "samples": {
    "inputs": {
      "sample_name": {
        "type": "string",
        "description": "Sample name"
      },
      "storage_cond": {
        "type": "string",
        "description": "Sample name"
      },
      "resource_volume": {
        "type": "volume",
        "description": "Volume of resource you would like to pipette."
      },
      "resource_id": {
        "type": "string",
        "description": "Resource id of resource you would like to provision, resource ids can be found here https://secure.transcriptic.com/_commercial/kits"
      }
    },
    "type": "group+",
    "description": "Samples you wish to create."
  }
}
```

evanworley commented on transcriptic/cli.py:
```diff
@@ -244,7 +244,8 @@ def upload_release(ctx, archive, package):
     help='Shows available protocols to be launched remotely'
 )
-def protocols(ctx, remote):
+@click.option("--json", "json_flag", help="print JSON repsponse", is_flag=True)
```
Typo "print JSON response"

evanworley commented on transcriptic/cli.py:
```diff
@@ -366,7 +369,8 @@ def delete_package(ctx, name, force):
 @click.pass_context
 @click.option("-i")
-def projects(ctx, i):
+@click.option("--json", "json_flag", help="print JSON repsponse", is_flag=True)
```
Typo here too

evanworley commented on transcriptic/cli.py:
```diff
@@ -395,7 +401,8 @@ def projects(ctx, i):
 @click.pass_context
 @click.argument('project_name')
-def runs(ctx, project_name):
+@click.option("--json", "json_flag", help="print JSON repsponse", is_flag=True)
```
Typo

evanworley commented on transcriptic/cli.py:
```diff
@@ -410,22 +417,32 @@ def runs(ctx, project_name):
+            for r in req:
+                run_list.append({'title': r['title'] or "(Untitled)",
+                                 'id': r['id'],
+                                 'completed_at': r['completed_at'].split("T")[0]
```
This and `created_at` might be more useful as time since epoch, since we are printing a machine readable format.

evanworley commented on transcriptic/cli.py:
```diff
@@ -410,22 +417,32 @@ def runs(ctx, project_name):
+                                 'completed_at': r['completed_at'].split("T")[0]
+                                   if r['completed_at'] else None,
+                                 'created_at': r['created_at'].split("T")[0],
+                                 'status': r['status'].replace("_", " ")})
```
Why do this replacement? JSON is intended to be processed by machines, not nice for humans to read

evanworley commented on transcriptic/cli.py:
```diff
@@ -410,22 +417,32 @@ def runs(ctx, project_name):
-                       u'{:^30}'.format(run[3]))
-            click.echo(u'{:-^120}'.format(''))
+        if json_flag:
+            run_list = []
```
Instead of accumulating in `run_list` you can just `map` `req`

bmiles commented:
Hey @yangchoo I updated this

bmiles commented:
@yangchoo here's the diff with master:

```json
diff --git a/CHANGELOG.md b/CHANGELOG.md
index 862e87a..50729c3 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,7 +1,10 @@
 ## `transcriptic` Changelog

 ## Unreleased
-Fixed
+
+## v3.12.0
+Added
+- New --json flag for runs, projects and protocols for fetching JSON
 - Fixed bug in _PlateRead that caused data overwrites if multiple instances of the same
 group_label were present

@@ -28,6 +31,10 @@ Fixed
 Updated
 - Removed setup.py requirement for `autoprotocol-python`

+## v3.9.3
+- Adds `--json` flag to cli commands: `projects`, `runs`, `protocols`. Provides
+raw JSON response.
+
 ## v3.9.2
 Fixed
 - Bugfix to resolve error caused by attempting to print unicode characters on the CLI.
diff --git a/transcriptic/cli.py b/transcriptic/cli.py
index 2c37dbb..4e9f19d 100755
--- a/transcriptic/cli.py
+++ b/transcriptic/cli.py
@@ -249,7 +249,8 @@ def upload_release(ctx, archive, package):
     default=False,
     help='Shows available protocols to be launched remotely'
 )
-def protocols(ctx, remote):
+@click.option("--json", "json_flag", help="print JSON response", is_flag=True)
+def protocols(ctx, remote, json_flag):
     """List protocols within your manifest or organization."""
     if remote:
         protocol_objs = ctx.obj.api.get_protocols()
@@ -260,15 +261,17 @@ def protocols(ctx, remote):
                        " is improperly formatted.")
             return
         protocol_objs = manifest['protocols']
-
-    click.echo('\n{:^60}'.format("Protocols within this {}:".format("organization" if remote else "manifest")))
-    click.echo('{:-^60}'.format(''))
-    for p in protocol_objs:
-        if p.get('display_name'):
-            display_str = u"{} ({})".format(p[u'name'], p.get(u'display_name'))
-        else:
-            display_str = p[u'name']
-        click.echo(u"{}\n{}".format(display_str, u'{:-^60}'.format("")))
+    if json_flag:
+        click.echo(json.dumps(protocol_objs))
+    else:
+        click.echo('\n{:^60}'.format("Protocols within this {}:".format("organization" if remote else "manifest")))
+        click.echo('{:-^60}'.format(''))
+        for p in protocol_objs:
+            if p.get('display_name'):
+                display_str = u"{} ({})".format(p[u'name'], p.get(u'display_name'))
+            else:
+                display_str = p[u'name']
+            click.echo(u"{}\n{}".format(display_str, u'{:-^60}'.format("")))


 @cli.command()
@@ -378,7 +381,8 @@ def delete_package(ctx, name, force):
 @cli.command()
 @click.pass_context
 @click.option("-i")
-def projects(ctx, i):
+@click.option("--json", "json_flag", help="print JSON response", is_flag=True)
+def projects(ctx, i, json_flag):
     """List the projects in your organization"""
     try:
         projects = ctx.obj.api.projects()
@@ -390,6 +394,8 @@ def projects(ctx, i):
             all_proj[proj["name"] + status] = proj["id"]
         if i:
             return proj_names
+        elif json_flag:
+            return click.echo(json.dumps(projects))
         else:
             click.echo('\n{:^80}'.format("PROJECTS:\n"))
             click.echo('{:^40}'.format("PROJECT NAME") + "|" +
@@ -407,7 +413,8 @@ def projects(ctx, i):
 @cli.command()
 @click.pass_context
 @click.argument('project_name')
-def runs(ctx, project_name):
+@click.option("--json", "json_flag", help="print JSON response", is_flag=True)
+def runs(ctx, project_name, json_flag):
     """List the runs that exist in a project"""
     project_id = get_project_id(project_name)
     run_list = []
@@ -422,22 +429,30 @@ def runs(ctx, project_name):
                              r['completed_at'].split("T")[0] if r['completed_at']
                              else r['created_at'].split("T")[0],
                              r['status'].replace("_", " ")])
-
-        click.echo(
-            '\n{:^120}'.format("Runs in Project '%s':\n" %
-                               get_project_name(project_name))
-        )
-        click.echo('{:^30}'.format("RUN TITLE") + "|" +
-                   '{:^30}'.format("RUN ID") + "|" +
-                   '{:^30}'.format("RUN DATE") + "|" +
-                   '{:^30}'.format('RUN STATUS'))
-        click.echo('{:-^120}'.format(''))
-        for run in run_list:
-            click.echo(u'{:^30}'.format(run[0]) + "|" +
-                       u'{:^30}'.format(run[1]) + "|" +
-                       u'{:^30}'.format(run[2]) + "|" +
-                       u'{:^30}'.format(run[3]))
-            click.echo(u'{:-^120}'.format(''))
+        if json_flag:
+            return click.echo(json.dumps(map(lambda x: {
+                                 'title': x['title'] or "(Untitled)",
+                                 'id': x['id'],
+                                 'completed_at': x['completed_at']
+                                 if x['completed_at'] else None,
+                                 'created_at': x['created_at'],
+                                 'status': x['status']}, req)))
+        else:
+            click.echo(
+                '\n{:^120}'.format("Runs in Project '%s':\n" %
+                                   get_project_name(project_name))
+            )
+            click.echo('{:^30}'.format("RUN TITLE") + "|" +
+                       '{:^30}'.format("RUN ID") + "|" +
+                       '{:^30}'.format("RUN DATE") + "|" +
+                       '{:^30}'.format('RUN STATUS'))
+            click.echo('{:-^120}'.format(''))
+            for run in run_list:
+                click.echo(u'{:^30}'.format(run[0]) + "|" +
+                           u'{:^30}'.format(run[1]) + "|" +
+                           u'{:^30}'.format(run[2]) + "|" +
+                           u'{:^30}'.format(run[3]))
+                click.echo(u'{:-^120}'.format(''))


 @cli.command("create-project")
diff --git a/transcriptic/version.py b/transcriptic/version.py
index fcbd695..d1a7f1e 100644
--- a/transcriptic/version.py
+++ b/transcriptic/version.py
@@ -1 +1 @@
-__version__ = "3.11.0"
+__version__ = "3.12.0"
```

yangchoo commented:
lgtm. thanks for posting the diff
